crates/api/src/wire.proto

```diff
@@ -0,0 +1,61 @@
+syntax = "proto3";
+
+package kitsune2.wire;
+
+// A Kitsune2 wire protocol message.
+//
+// This is the top-level encoding
+// that will be transferred between Kitsune2 peers. Most communications
+// between peers to make Kitsune2 actually function will be encoded
+// separately inside the payload of TY_MODULE type messages.
+message K2Proto {
+  // Enumeration of the types of messages that can be sent between peers.
+  //
+  // We are using this enum field to distinguish between top-level messages,
+  // rather than protobuf's oneof because of the downsides of upgrading oneofs.
+  enum Ty {
+    // The "UNSPECIFIED" type for future message types.
+    //
+    // In general, peers should ignore unspecified messages, but
+    // should still count them toward any ratelimiting metrics.
+    TY_UNSPECIFIED = 0;
+
+    // This message is preflight data.
+    //
+    // The implementor is responsible for encoding any module-specific
+    // requirements within the data payload of this message type.
+    // For example, if peers are required to include the "dht_v1" module
+    // to communicate with each other, they should reject preflight to
+    // peers that do not include that module.
+    TY_PREFLIGHT = 1;
+
+    // This is a notification or fire-and-forget message from a peer.
+    //
+    // This type requires that a "space" be specified.
+    TY_NOTIFY = 2;
+
+    // This is a module communication.
+    //
+    // This type requires that a "space" be specified.
+    // This type requires that a "module" be specified.
+    TY_MODULE = 3;
+
+    // This message indicates a general disconnect, with the reason
+    // or context specified in the data payload.
+    //
+    // We may add additional specific disconnect codes in the future.
+    TY_DISCONNECT_GENERAL = 900;
+  }
+
+  // The type of this message.
+  optional Ty ty = 1;
+
+  // The payload or content of this message.
+  optional bytes data = 2;
+
+  // If the Ty requires that a space be specified, this is it.
+  optional bytes space = 3;
+
+  // If the Ty requires a module impl be specified, this is it.
```

Comment:
Might be good to mention some modules as examples, so it's clear what this field is for? Module is a fairly general term but we have a specific meaning in mind here

Reply:
Done: https://github.com/holochain/kitsune2/pull/38/commits/522892f79d9840fd718edc066bb69b8343a20422